Updates to Avro type section & remove experimental

kathancox opened the pull request:
Closes #10881 

PR includes the following updates: 

- Adjusted the Avro Limitations section in 21.2 to remove the decimal bullets that no longer apply and detail on `BIT` + `VARBIT` type encoding, and added an example on how the array encoding works.
- Adjusted same section in v21.1 & 20.2 per the details provided in the issue. 
- Edited the Avro type table for 21.2 & 21.1 (besides `DECIMAL`)... Note: Left the `DECIMAL` type in the table unchanged in 21.1.
-  Removed `experimental` from `experimental-avro` across CDC docs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/11066)
<!-- Reviewable:end -->

amruss commented on v20.2/create-changefeed.md:
```diff
@@ -109,15 +109,15 @@ Option | Value | Description
+You can only emit an Avro record if you are using a changefeed [connected to Kafka](#create-a-changefeed-connected-to-kafka-using-avro). Below is a list of unsupported SQL types and values for Avro changefeeds:
 
 - [Decimals](decimal.html) must have precision specified.
 - [Decimals](decimal.html) with `NaN` or infinite values cannot be written in Avro.
```
This no longer applies, we should take it out as well

amruss commented on v21.1/create-changefeed.md:
```diff
@@ -129,7 +127,17 @@ Below is a list of unsupported SQL types and values for Avro changefeeds:
+
+  For efficiency, CockroachDB encodes `BIT` and `VARBIT` bitfield types as arrays of 64-bit integers. Encoding in CockroachDB is [big-endian](https://en.wikipedia.org/wiki/Endianness), therefore the last value may have many trailing zeroes. For this reason, the first value of each array is the number of bits that are used in the last value of the array.
+
+  For instance, if the bitfield is 129 bits long, there will be 4 integers in the array. The first integer will be `1`; representing the number of bits in the last value, the second integer will be the first 64 bits, the third integer will be bits 65–128, and the last integer will either be `0` or `9223372036854775808` (i.e. the integer with only the first bit set).
```
It might make it more clear to show what `9223372036854775808` encodes to

amruss replied:
and that we're translating base 2 encoding -> base 10 encoding, and the user needs to do the opposite downstream: take the base 10 encoded number and base 2 encode it

amruss reviewed (changes requested):
Couple of small things

kathancox commented:
Hey @amruss, True! I have added more detail into the `BIT` + `VARBIT` explanation. I added another example as part of showing the conversion on the user's end — I felt like two different sets of numbers/arrays would demonstrate how this works out a little better, but open to editing that!

HonoreDB commented on v21.1/create-changefeed.md:
```diff
@@ -129,7 +127,33 @@ Below is a list of unsupported SQL types and values for Avro changefeeds:
+
+    For downstream processing, it is necessary to base 2 encode every element in the array (except for the first element). The first number in the array gives you the number of bits to take from the last base 2 number — that is, the most significant bits. So, in the example above this would be `1`. Finally, all the base 2 numbers can be appended together, which will result in the original number  ...
+
+    To clarify this process, in another example in which you had a bitfield of 135 bits, the array would be as follow when base 10:
```
If you're going to have a second example, I think it's a little clearer if the last used digit is a 0, to make it clear that you can't just eliminate all trailing zeroes from the last value. Like if you changed this from 135 bits to 136 bits, so that we're taking 8 bits from the last value and getting `10111110`.

kathancox replied:
Yes, that's a really good point

HonoreDB reviewed (commented):
Thanks again, especially for the worked examples of BIT encoding.

kathancox commented on _includes/v21.1/cdc/create-core-changefeed-avro.md:
```diff
@@ -57,7 +57,7 @@ In this example, you'll set up a core changefeed for a single-node cluster that
     {% include copy-clipboard.html %}
     ~~~ sql
-    > EXPERIMENTAL CHANGEFEED FOR bar WITH format = experimental_avro, confluent_schema_registry = 'http://localhost:8081';
+    > EXPERIMENTAL CHANGEFEED FOR bar WITH format = avro, confluent_schema_registry = 'http://localhost:8081';
```
Ack, OK. @amruss so I just want to remove `experimental` from avro for v21.2?

amruss replied:
I think lets just keep it in 21.2

amruss commented on _includes/v20.2/cdc/create-core-changefeed-avro.md:
```diff
@@ -57,7 +57,7 @@ In this example, you'll set up a core changefeed for a single-node cluster that
     {% include copy-clipboard.html %}
     ~~~ sql
-    > EXPERIMENTAL CHANGEFEED FOR bar WITH format = experimental_avro, confluent_schema_registry = 'http://localhost:8081';
+    > EXPERIMENTAL CHANGEFEED FOR bar WITH format = avro, confluent_schema_registry = 'http://localhost:8081';
```
I think you're right, I think just keeping it non-experimental in 21.2 is fine

amruss reviewed (approved):
LGTM, assuming the core piece is also updated

kathancox commented:
> LGTM, assuming the core piece is also updated

Thanks @amruss I have incorporated the updated avro limitation section into the core changefeed docs. Would welcome a check of that, but I'll assign the docs reviewer too.

kathancox commented:
cc @mikeCRL

HonoreDB reviewed (approved):
> Reviewed 2 of 9 files at r1, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/11066)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @kathancox, and @rmloveland)



<!-- Sent from Reviewable.io -->

rmloveland commented on v21.1/changefeed-for.md:
```diff
@@ -57,16 +57,38 @@ Option | Value | Description
+    {"array": [1, <first 64 bits>, <second 64 bits>, 0 or 9223372036854775808]}
+    ~~~
+
+    For downstream processing, it is necessary to base 2 encode every element in the array (except for the first element). The first number in the array gives you the number of bits to take from the last base-2 number — that is, the most significant bits. So, in the example above this would be `1`. Finally, all the base-2 numbers can be appended together, which will result in the original number  ...
```
we should say "base-2" to match our base-* language elsewhere

rmloveland commented on v21.1/changefeed-for.md:
```diff
@@ -57,16 +57,38 @@ Option | Value | Description
+    [8, 1111110111011011111111111111111111111111111111111111111111111111, 1111111111111111111111111111111111111111111111111111111111111111, 1011111000000000000000000000000000000000000000000000000000000000]
+    ~~~
+
+    Next, you use the first number to take the number of bits from the last base-2 element, `10111110`. Finally, you append each of the base-2 numbers (the last number from the array being truncated), which in this case results in 136 bits, the original number of bits.
```
> Finally, you append each of the base-2 numbers (the last number from the array being truncated)

I think I get it, but the parenthetical makes the ordering feel reversed.  After reading this a few times I understand this to mean that I need to snarf the # of bits in array element 1 from the 4th element of the array to give some number _n_, and then throw away (drop) that 4th element.  Then, append array elements 2 and 3 to _n_, making one big 136-bit binary number?

may be worth revising slightly so the order is more explicit. then again I think I was able to figure it out (?) from what you wrote, so this is probably an optional comment.  And also this is not my area, so the fact I was a bit confused is not a showstopper

rmloveland commented on v21.2/changefeed-for.md:
```diff
@@ -52,21 +52,41 @@ Option | Value | Description
+    For efficiency, CockroachDB encodes `BIT` and `VARBIT` bitfield types as arrays of 64-bit integers. That is, [base-2 (binary format)](https://en.wikipedia.org/wiki/Binary_number#Conversion_to_and_from_other_numeral_systems) `BIT` and `VARBIT` data types are converted to base 10 and stored in arrays. Encoding in CockroachDB is [big-endian](https://en.wikipedia.org/wiki/Endianness), therefore t ...
 
-- [`time`, `date`, `interval`](https://github.com/cockroachdb/cockroach/issues/32472), [`uuid`, `inet`](https://github.com/cockroachdb/cockroach/issues/34417), [`array`](https://github.com/cockroachdb/cockroach/issues/34420), and [`jsonb`](https://github.com/cockroachdb/cockroach/issues/34421) are not supported in Avro yet.
+    For instance, if the bitfield is 129 bits long, there will be 4 integers in the array. The first integer will be `1`; representing the number of bits in the last value, the second integer will be the first 64 bits, the third integer will be bits 65–128, and the last integer will either be `0` or `9223372036854775808` (i.e. the integer with only the first bit set, or `1000000000000000000000000 ...
```
This section is repeated a bit, could maybe be an include?

rmloveland commented on v21.2/create-changefeed.md:
```diff
@@ -100,36 +100,54 @@ Option | Value | Description
 - [Decimals](decimal.html) must have precision specified.
-- [Decimals](decimal.html) with `NaN` or infinite values cannot be written in Avro.
+- [`BIT`](bit.html) and [`VARBIT`](bit.html) types are encoded as arrays of 64-bit integers.
+
```
Definitely feeling an include++

rmloveland commented on v21.2/create-changefeed.md:
```diff
@@ -145,10 +163,20 @@ CockroachDB Type | Avro Type | Avro Logical Type
 [`UUID`](uuid.html) | [`STRING`](https://avro.apache.org/docs/1.8.1/spec.html#schema_primitive) |
 [`INET`](inet.html) | [`STRING`](https://avro.apache.org/docs/1.8.1/spec.html#schema_primitive) |
 [`JSONB`](jsonb.html) | [`STRING`](https://avro.apache.org/docs/1.8.1/spec.html#schema_primitive) |
+[`ENUMS`](enum.html) | [`STRING`](https://avro.apache.org/docs/1.8.1/spec.html#schema_primitive) |
```
Is this list of Avro supported types the same as on the other page?  If so, recommend an include

kathancox replied:
At least two of the versions have slightly different availability

rmloveland reviewed (approved):
LGTM with a few comments which are mostly take-or-leavable
